File: isl-transform/src/main/kotlin/com/intuit/isl/runtime/TransformCompiler.kt
package com.intuit.isl.runtime

import com.intuit.isl.commands.builder.ExecutionBuilder
import com.intuit.isl.common.IOperationContext
import com.intuit.isl.parser.TransformParser

class TransformCompiler {
    /**
     * Compile a full ISL Script (includes functions, modifiers, imports)
     */
    fun compileIsl(
        moduleName: String, script: String,
        moduleFinder: ((name: String) -> ITransformer?)? = null
    ): ITransformer {
        val moduleToken = TransformParser().parseTransform(moduleName, script);

        val module = ExecutionBuilder(moduleName, moduleToken, moduleFinder).build();

        val transformer = Transformer(module);

        return transformer;
    }

    fun compileLocalIsl(
        moduleName: String, script: String,
        localContext: IOperationContext
    ): ILocalTransformer {
        val moduleToken = TransformParser().parseTransform(moduleName, script);

        val module = ExecutionBuilder(moduleName, moduleToken, null, localContext).build();

        val transformer = LocalTransformer(module);
        return transformer;
    }


}

File: isl-csharp/Isl/TransformCompiler.cs
using Isl.Compiler;
using Isl.Parser;
using Isl.Runtime;

namespace Isl;

/// <summary>
/// Public compile entry point. Mirrors Kotlin's <c>TransformCompiler.compileIsl</c>:
/// lex + parse + lower to commands + return a runnable <see cref="ITransformer"/>.
/// </summary>
public static class TransformCompiler
{
    public static ITransformer CompileIsl(string name, string script)
    {
        var lexer = new IslLexer(script);
        var tokens = lexer.Tokenize();
        var parser = new IslParser(tokens);
        var module = parser.ParseModule();
        var execModule = new ExecutionBuilder(name, module).Build();
        return new Transformer(execModule);
    }
}
